src/shims/posix/linux/foreign_items.rs

```diff
@@ -192,3 +201,61 @@ fn getrandom<'tcx>(
     this.write_scalar(Scalar::from_machine_usize(len, this), dest)?;
     Ok(())
 }
+
+fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    if args.len() < 4 {
+        throw_ub_format!("incorrect number of arguments for futex syscall: got {}, expected at least 4", args.len());
+    }
+    let addr = this.read_scalar(args[1])?.check_init()?;
+    let op = this.read_scalar(args[2])?.to_i32()?;
+    let val = this.read_scalar(args[3])?.to_i32()?;
+
+    this.memory.check_ptr_access(addr, Size::from_bytes(4), Align::from_bytes(4).unwrap())?;
+
+    let addr = addr.assert_ptr();
+
+    let thread = this.get_active_thread();
+
+    let futex_private = this.eval_libc_i32("FUTEX_PRIVATE_FLAG")?;
+    let futex_wait = this.eval_libc_i32("FUTEX_WAIT")?;
+    let futex_wake = this.eval_libc_i32("FUTEX_WAKE")?;
+
+    match op & !futex_private {
+        op if op == futex_wait => {
+            if args.len() < 5 {
+                throw_ub_format!("incorrect number of arguments for FUTEX_WAIT syscall: got {}, expected at least 5", args.len());
+            }
+            let timeout = this.read_scalar(args[4])?.check_init()?;
+            if !this.is_null(timeout)? {
+                throw_ub_format!("miri does not support timeouts for futex operations");
+            }
+            let futex_val = this.read_scalar_at_offset(args[1], 0, this.machine.layouts.i32)?.to_i32()?;
```

Comment:
What is happening here, why is `args[1]` used again? Usually the arguments should just be read once at the top, and that's it.

Also why is this ptr-sized value (?) now treated as a`i32`?

Reply:
`args[1]` should be a pointer to a (properly aligned) (atomic) i32. futex_wait needs the value of that i32, to compare it with `args[3]`, which is a i32.

(And since this uses the varargs syscall() function, there's no strict signature. Callers might pass that pointer as a `*const AtomicI32` or `*mut i32` or `*const u32` etc.)

Reply:
Can we rely on it being some ptr type to a 32bit integer scalar? Then you can use `deref_operand` and do not have to repeat the layout here.

Reply:
When I did that, the test failed with an error about being unable to read AtomicI32. I guess because a `*const AtomicI32` was passed instead of a `*const i32`:

```rust
let futex_val = this.read_scalar(this.deref_operand(args[1])?.into())?.to_i32()?;
```

```
"message":"rust/compiler/rustc_mir/src/interpret/operand.rs:327:13: primitive read failed for type: std::sync::atomic::AtomicI32"
```

Reply:
Strange, not sure why that type doesn't behave like a scalar.

But either way, `args[1]` should be used only exactly once, in a statement like
```rust
let futex_val_place = this.deref_operand(args[1])?;
```
Then later you can use `this.read_scalar(futex_val_place.offset)` to read the actual data, and `futex_val_place.ptr.assert_ptr()` to get the `Pointer`. (`deref_operand` does the necessary int-to-ptr conversion.)

Reply:
Ah, thanks. That helps a lot.